File: ainpc-core-plugin/src/main/java/ro/ainpc/spawn/HouseAllocation.java
package ro.ainpc.spawn;

import ro.ainpc.npc.AINPC;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

public record HouseAllocation(
    String placeId,
    String familyId,
    String primaryOwnerNpcKey,
    int maxResidents,
    List<ResidentPlan> residentPlans
) {
    public HouseAllocation {
        placeId = clean(placeId);
        familyId = clean(familyId);
        primaryOwnerNpcKey = clean(primaryOwnerNpcKey);
        residentPlans = List.copyOf(residentPlans != null ? residentPlans : List.of());
        maxResidents = maxResidents > 0 ? maxResidents : residentPlans.size();
    }

    public static Builder builder(String placeId) {
        return new Builder(placeId);
    }

    public List<NpcSpawnPlan> toNpcSpawnPlans() {
        List<NpcSpawnPlan> plans = new ArrayList<>();
        for (ResidentPlan resident : residentPlans) {
            plans.add(resident.toNpcSpawnPlan(placeId, familyId));
        }
        return List.copyOf(plans);
    }

    public Map<String, String> toPlaceMetadata() {
        Map<String, String> metadata = new LinkedHashMap<>();
        metadata.put("role", "home");
        if (!familyId.isBlank()) {
            metadata.put("family_id", familyId);
        }
        metadata.put("max_residents", Integer.toString(maxResidents));
        metadata.put("residents", String.join(",", residentNpcKeys()));
        return Map.copyOf(metadata);
    }

    public FamilyBindingPlan toFamilyBindingPlan(Map<String, AINPC> spawnedNpcsByKey) {
        List<FamilyBindingPlan.Member> members = new ArrayList<>();
        Map<String, AINPC> resolvedNpcs = spawnedNpcsByKey != null ? spawnedNpcsByKey : Map.of();

        for (ResidentPlan resident : residentPlans) {
            AINPC npc = resolvedNpcs.get(normalizeKey(resident.npcKey()));
            if (npc == null) {
                npc = resolvedNpcs.get(resident.npcKey());
            }
            members.add(FamilyBindingPlan.member(resident.npcKey(), npc, resident.relationRole()));
        }

        return new FamilyBindingPlan(familyId, members);
    }

    public List<String> residentNpcKeys() {
        return residentPlans.stream()
            .map(ResidentPlan::npcKey)
            .filter(key -> !key.isBlank())
            .toList();
    }

    public static final class Builder {
        private final String placeId;
        private String familyId = "";
        private String primaryOwnerNpcKey = "";
        private int maxResidents = 0;
        private final List<ResidentPlan> residentPlans = new ArrayList<>();

        private Builder(String placeId) {
            this.placeId = placeId;
        }

        public Builder familyId(String familyId) {
            this.familyId = familyId;
            return this;
        }

        public Builder primaryOwnerNpcKey(String primaryOwnerNpcKey) {
            this.primaryOwnerNpcKey = primaryOwnerNpcKey;
            return this;
        }

        public Builder maxResidents(int maxResidents) {
            this.maxResidents = maxResidents;
            return this;
        }

        public Builder addResident(ResidentPlan residentPlan) {
            if (residentPlan != null) {
                this.residentPlans.add(residentPlan);
            }
            return this;
        }

        public Builder residents(List<ResidentPlan> residentPlans) {
            this.residentPlans.clear();
            if (residentPlans != null) {
                this.residentPlans.addAll(residentPlans);
            }
            return this;
        }

        public HouseAllocation build() {
            return new HouseAllocation(placeId, familyId, primaryOwnerNpcKey, maxResidents, residentPlans);
        }
    }

    public record ResidentPlan(
        String npcKey,
        String name,
        String relationRole,
        String occupation,
        String backstory,
        int age,
        String gender,
        String archetype,
        String spawnNodeId,
        String homeNodeId,
        String bedNodeId,
        String workPlaceId,
        String workNodeId,
        String socialPlaceId,
        String socialNodeId
    ) {
        public ResidentPlan {
            npcKey = clean(npcKey);
            name = clean(name);
            relationRole = clean(relationRole);
            occupation = clean(occupation);
            backstory = clean(backstory);
            age = age > 0 ? age : 30;
            gender = clean(gender);
            archetype = clean(archetype);
            spawnNodeId = clean(spawnNodeId);
            homeNodeId = clean(homeNodeId);
            bedNodeId = clean(bedNodeId);
            workPlaceId = clean(workPlaceId);
            workNodeId = clean(workNodeId);
            socialPlaceId = clean(socialPlaceId);
            socialNodeId = clean(socialNodeId);
        }

        public static ResidentBuilder builder(String npcKey, String name) {
            return new ResidentBuilder(npcKey, name);
        }

        public String effectiveHomeNodeId() {
            return !homeNodeId.isBlank() ? homeNodeId : bedNodeId;
        }

        public NpcSpawnPlan toNpcSpawnPlan(String homePlaceId, String familyId) {
            return NpcSpawnPlan.builder(npcKey, !name.isBlank() ? name : humanizeId(npcKey))
                .occupation(occupation)
                .backstory(backstory)
                .age(age)
                .gender(gender)
                .archetype(archetype)
                .homePlaceId(homePlaceId)
                .workPlaceId(workPlaceId)
                .socialPlaceId(socialPlaceId)
                .spawnNodeId(spawnNodeId)
                .homeNodeId(effectiveHomeNodeId())
                .workNodeId(workNodeId)
                .socialNodeId(socialNodeId)
                .familyId(familyId)
                .build();
        }
    }

    public static final class ResidentBuilder {
        private final String npcKey;
        private final String name;
        private String relationRole = "";
        private String occupation = "";
        private String backstory = "";
        private int age = 30;
        private String gender = "male";
        private String archetype = "";
        private String spawnNodeId = "";
        private String homeNodeId = "";
        private String bedNodeId = "";
        private String workPlaceId = "";
        private String workNodeId = "";
        private String socialPlaceId = "";
        private String socialNodeId = "";

        private ResidentBuilder(String npcKey, String name) {
            this.npcKey = npcKey;
            this.name = name;
        }

        public ResidentBuilder relationRole(String relationRole) {
            this.relationRole = relationRole;
            return this;
        }

        public ResidentBuilder occupation(String occupation) {
            this.occupation = occupation;
            return this;
        }

        public ResidentBuilder backstory(String backstory) {
            this.backstory = backstory;
            return this;
        }

        public ResidentBuilder age(int age) {
            this.age = age;
            return this;
        }

        public ResidentBuilder gender(String gender) {
            this.gender = gender;
            return this;
        }

        public ResidentBuilder archetype(String archetype) {
            this.archetype = archetype;
            return this;
        }

        public ResidentBuilder spawnNodeId(String spawnNodeId) {
            this.spawnNodeId = spawnNodeId;
            return this;
        }

        public ResidentBuilder homeNodeId(String homeNodeId) {
            this.homeNodeId = homeNodeId;
            return this;
        }

        public ResidentBuilder bedNodeId(String bedNodeId) {
            this.bedNodeId = bedNodeId;
            return this;
        }

        public ResidentBuilder workPlaceId(String workPlaceId) {
            this.workPlaceId = workPlaceId;
            return this;
        }

        public ResidentBuilder workNodeId(String workNodeId) {
            this.workNodeId = workNodeId;
            return this;
        }

        public ResidentBuilder socialPlaceId(String socialPlaceId) {
            this.socialPlaceId = socialPlaceId;
            return this;
        }

        public ResidentBuilder socialNodeId(String socialNodeId) {
            this.socialNodeId = socialNodeId;
            return this;
        }

        public HouseAllocation.ResidentPlan build() {
            return new HouseAllocation.ResidentPlan(
                npcKey,
                name,
                relationRole,
                occupation,
                backstory,
                age,
                gender,
                archetype,
                spawnNodeId,
                homeNodeId,
                bedNodeId,
                workPlaceId,
                workNodeId,
                socialPlaceId,
                socialNodeId
            );
        }
    }

    private static String clean(String value) {
        return value == null ? "" : value.trim();
    }

    private static String normalizeKey(String value) {
        return clean(value).toLowerCase(Locale.ROOT).replace(' ', '_').replace('-', '_');
    }

    private static String humanizeId(String rawId) {
        String value = clean(rawId);
        int scopeSeparator = value.lastIndexOf(':');
        if (scopeSeparator >= 0) {
            value = value.substring(scopeSeparator + 1);
        }

        StringBuilder builder = new StringBuilder();
        for (String part : value.split("[_-]+")) {
            if (part.isBlank()) {
                continue;
            }
            if (builder.length() > 0) {
                builder.append(' ');
            }
            builder.append(Character.toUpperCase(part.charAt(0)));
            if (part.length() > 1) {
                builder.append(part.substring(1));
            }
        }
        return builder.length() > 0 ? builder.toString() : value;
    }
}

File: ainpc-core-plugin/bin/main/ro/ainpc/spawn/HouseAllocation.kt
package ro.ainpc.spawn

import ro.ainpc.npc.AINPC
import java.util.LinkedHashMap
import java.util.Locale

data class HouseAllocation(
    val placeId: String,
    val familyId: String,
    val primaryOwnerNpcKey: String,
    val maxResidents: Int,
    val residentPlans: List<ResidentPlan>
) {
    private val cleanedPlaceId = clean(placeId)
    private val cleanedFamilyId = clean(familyId)
    private val cleanedPrimaryOwnerNpcKey = clean(primaryOwnerNpcKey)
    private val cleanedResidentPlans = residentPlans.toList()
    private val effectiveMaxResidents = if (maxResidents > 0) maxResidents else cleanedResidentPlans.size

    fun placeId(): String = cleanedPlaceId
    fun familyId(): String = cleanedFamilyId
    fun primaryOwnerNpcKey(): String = cleanedPrimaryOwnerNpcKey
    fun maxResidents(): Int = effectiveMaxResidents
    fun residentPlans(): List<ResidentPlan> = cleanedResidentPlans.toList()

    fun toNpcSpawnPlans(): List<NpcSpawnPlan> {
        val plans = mutableListOf<NpcSpawnPlan>()
        for (resident in cleanedResidentPlans) {
            plans.add(resident.toNpcSpawnPlan(cleanedPlaceId, cleanedFamilyId))
        }
        return plans.toList()
    }

    fun householdId(): String {
        if (cleanedFamilyId.isNotBlank()) {
            return cleanedFamilyId
        }
        val normalizedPlaceId = normalizeKey(cleanedPlaceId)
        return if (normalizedPlaceId.isBlank()) "" else "household_$normalizedPlaceId"
    }

    fun toPlaceMetadata(): Map<String, String> {
        val metadata = LinkedHashMap<String, String>()
        metadata["role"] = "home"
        if (cleanedFamilyId.isNotBlank()) {
            metadata["family_id"] = cleanedFamilyId
        }
        metadata["max_residents"] = effectiveMaxResidents.toString()
        metadata["residents"] = residentNpcKeys().joinToString(",")
        return metadata.toMap()
    }

    fun toFamilyBindingPlan(spawnedNpcsByKey: Map<String, AINPC>?): FamilyBindingPlan {
        val members = mutableListOf<FamilyBindingPlan.Member>()
        val resolvedNpcs = spawnedNpcsByKey ?: mapOf()

        for (resident in cleanedResidentPlans) {
            var npc = resolvedNpcs[normalizeKey(resident.npcKey())]
            if (npc == null) {
                npc = resolvedNpcs[resident.npcKey()]
            }
            members.add(FamilyBindingPlan.member(resident.npcKey(), npc, resident.relationRole()))
        }

        return FamilyBindingPlan(cleanedFamilyId, members)
    }

    fun residentNpcKeys(): List<String> {
        return cleanedResidentPlans
            .map { it.npcKey() }
            .filter { key -> key.isNotBlank() }
    }

    class Builder private constructor(private val placeId: String) {
        private var familyId = ""
        private var primaryOwnerNpcKey = ""
        private var maxResidents = 0
        private val residentPlans = mutableListOf<ResidentPlan>()

        fun familyId(familyId: String): Builder {
            this.familyId = familyId
            return this
        }

        fun primaryOwnerNpcKey(primaryOwnerNpcKey: String): Builder {
            this.primaryOwnerNpcKey = primaryOwnerNpcKey
            return this
        }

        fun maxResidents(maxResidents: Int): Builder {
            this.maxResidents = maxResidents
            return this
        }

        fun addResident(residentPlan: ResidentPlan?): Builder {
            if (residentPlan != null) {
                residentPlans.add(residentPlan)
            }
            return this
        }

        fun residents(residentPlans: List<ResidentPlan>?): Builder {
            this.residentPlans.clear()
            if (residentPlans != null) {
                this.residentPlans.addAll(residentPlans)
            }
            return this
        }

        fun build(): HouseAllocation {
            return HouseAllocation(placeId, familyId, primaryOwnerNpcKey, maxResidents, residentPlans)
        }

        companion object {
            @JvmStatic
            fun create(placeId: String): Builder = Builder(placeId)
        }
    }

    data class ResidentPlan(
        val npcKey: String,
        val name: String,
        val relationRole: String,
        val occupation: String,
        val backstory: String,
        val age: Int,
        val gender: String,
        val archetype: String,
        val spawnNodeId: String,
        val homeNodeId: String,
        val bedNodeId: String,
        val workPlaceId: String,
        val workNodeId: String,
        val socialPlaceId: String,
        val socialNodeId: String
    ) {
        private val cleanedNpcKey = clean(npcKey)
        private val cleanedName = clean(name)
        private val cleanedRelationRole = clean(relationRole)
        private val cleanedOccupation = clean(occupation)
        private val cleanedBackstory = clean(backstory)
        private val cleanedAge = if (age > 0) age else 30
        private val cleanedGender = clean(gender)
        private val cleanedArchetype = clean(archetype)
        private val cleanedSpawnNodeId = clean(spawnNodeId)
        private val cleanedHomeNodeId = clean(homeNodeId)
        private val cleanedBedNodeId = clean(bedNodeId)
        private val cleanedWorkPlaceId = clean(workPlaceId)
        private val cleanedWorkNodeId = clean(workNodeId)
        private val cleanedSocialPlaceId = clean(socialPlaceId)
        private val cleanedSocialNodeId = clean(socialNodeId)

        fun npcKey(): String = cleanedNpcKey
        fun name(): String = cleanedName
        fun relationRole(): String = cleanedRelationRole
        fun occupation(): String = cleanedOccupation
        fun backstory(): String = cleanedBackstory
        fun age(): Int = cleanedAge
        fun gender(): String = cleanedGender
        fun archetype(): String = cleanedArchetype
        fun spawnNodeId(): String = cleanedSpawnNodeId
        fun homeNodeId(): String = cleanedHomeNodeId
        fun bedNodeId(): String = cleanedBedNodeId
        fun workPlaceId(): String = cleanedWorkPlaceId
        fun workNodeId(): String = cleanedWorkNodeId
        fun socialPlaceId(): String = cleanedSocialPlaceId
        fun socialNodeId(): String = cleanedSocialNodeId

        fun effectiveHomeNodeId(): String = if (cleanedHomeNodeId.isNotBlank()) cleanedHomeNodeId else cleanedBedNodeId

        fun toNpcSpawnPlan(homePlaceId: String, familyId: String): NpcSpawnPlan {
            return NpcSpawnPlan.builder(cleanedNpcKey, if (cleanedName.isNotBlank()) cleanedName else humanizeId(cleanedNpcKey))
                .occupation(cleanedOccupation)
                .backstory(cleanedBackstory)
                .age(cleanedAge)
                .gender(cleanedGender)
                .archetype(cleanedArchetype)
                .homePlaceId(homePlaceId)
                .workPlaceId(cleanedWorkPlaceId)
                .socialPlaceId(cleanedSocialPlaceId)
                .spawnNodeId(cleanedSpawnNodeId)
                .homeNodeId(effectiveHomeNodeId())
                .workNodeId(cleanedWorkNodeId)
                .socialNodeId(cleanedSocialNodeId)
                .familyId(familyId)
                .build()
        }

        companion object {
            @JvmStatic
            fun builder(npcKey: String, name: String): ResidentBuilder = ResidentBuilder.create(npcKey, name)
        }
    }

    class ResidentBuilder private constructor(
        private val npcKey: String,
        private val name: String
    ) {
        private var relationRole = ""
        private var occupation = ""
        private var backstory = ""
        private var age = 30
        private var gender = "male"
        private var archetype = ""
        private var spawnNodeId = ""
        private var homeNodeId = ""
        private var bedNodeId = ""
        private var workPlaceId = ""
        private var workNodeId = ""
        private var socialPlaceId = ""
        private var socialNodeId = ""

        fun relationRole(relationRole: String): ResidentBuilder {
            this.relationRole = relationRole
            return this
        }

        fun occupation(occupation: String): ResidentBuilder {
            this.occupation = occupation
            return this
        }

        fun backstory(backstory: String): ResidentBuilder {
            this.backstory = backstory
            return this
        }

        fun age(age: Int): ResidentBuilder {
            this.age = age
            return this
        }

        fun gender(gender: String): ResidentBuilder {
            this.gender = gender
            return this
        }

        fun archetype(archetype: String): ResidentBuilder {
            this.archetype = archetype
            return this
        }

        fun spawnNodeId(spawnNodeId: String): ResidentBuilder {
            this.spawnNodeId = spawnNodeId
            return this
        }

        fun homeNodeId(homeNodeId: String): ResidentBuilder {
            this.homeNodeId = homeNodeId
            return this
        }

        fun bedNodeId(bedNodeId: String): ResidentBuilder {
            this.bedNodeId = bedNodeId
            return this
        }

        fun workPlaceId(workPlaceId: String): ResidentBuilder {
            this.workPlaceId = workPlaceId
            return this
        }

        fun workNodeId(workNodeId: String): ResidentBuilder {
            this.workNodeId = workNodeId
            return this
        }

        fun socialPlaceId(socialPlaceId: String): ResidentBuilder {
            this.socialPlaceId = socialPlaceId
            return this
        }

        fun socialNodeId(socialNodeId: String): ResidentBuilder {
            this.socialNodeId = socialNodeId
            return this
        }

        fun build(): ResidentPlan {
            return ResidentPlan(
                npcKey,
                name,
                relationRole,
                occupation,
                backstory,
                age,
                gender,
                archetype,
                spawnNodeId,
                homeNodeId,
                bedNodeId,
                workPlaceId,
                workNodeId,
                socialPlaceId,
                socialNodeId
            )
        }

        companion object {
            @JvmStatic
            fun create(npcKey: String, name: String): ResidentBuilder = ResidentBuilder(npcKey, name)
        }
    }

    companion object {
        @JvmStatic
        fun builder(placeId: String): Builder = Builder.create(placeId)

        private fun clean(value: String?): String = value?.trim() ?: ""

        private fun normalizeKey(value: String): String {
            return clean(value).lowercase(Locale.ROOT).replace(' ', '_').replace('-', '_')
        }

        private fun humanizeId(rawId: String): String {
            var value = clean(rawId)
            val scopeSeparator = value.lastIndexOf(':')
            if (scopeSeparator >= 0) {
                value = value.substring(scopeSeparator + 1)
            }

            val builder = StringBuilder()
            for (part in value.split(Regex("[_-]+"))) {
                if (part.isBlank()) {
                    continue
                }
                if (builder.isNotEmpty()) {
                    builder.append(' ')
                }
                builder.append(part[0].uppercaseChar())
                if (part.length > 1) {
                    builder.append(part.substring(1))
                }
            }
            return if (builder.isNotEmpty()) builder.toString() else value
        }
    }
}
